compiler/mir/mirtrees.nim

```diff
@@ -56,166 +56,138 @@ type
-                 ## location may or may not contain a value and is not
-                 ## responsible for destroying it
-                 ## inputs: either a single value (no arg-block) or none
-    mnkDefUnpack ## starts the lifetime of a location that owns a *tuple*
-                 ## value but is not responsible for destroying it, as all
-                 ## sub-values are moved out of it
-                 ## **Warning**: this is a hack and it should be relied on as
-                 ## little as possible. It's a workaround to support the
-                 ## temporary tuple values introduced for destructuring.
-
-    mnkFastAsgn ## ``fastAsgn(dst, src)``; assigns the `src` value to the location
-                ## named by the lvalue `dst`. Neither the previous value in the
-                ## destination location nor the source value are mutated in any
-                ## way. No transfer of ownership happens.
-    mnkAsgn     ## ``asgn(dst, src)``; assigns the `src` value to the location
-                ##  named by `dst`, also transferring onwership.
-    mnkInit     ## ``init(dst, src)``; similar to `asgn`, but with the
-                ## guarantee that the destination contains no value prior
-
-    mnkSwitch ## ``switch(x, y)``; changes the active branch of the record-case
-              ## identified by the result of the ``pathVariant`` operation used
-              ## as the `x` operand. `y` is the new discriminator value
+    # future direction:
+    # store the type of the destination within each def, assignment, etc. and
+    # then remove the type field from ``MirNode``
+
+    mnkDef       ## marks the start of existence of a local, global, procedure,
+                 ## or temporary. Supports an optional intial value (except for
+                 ## procedure definitions)
+    mnkDefCursor ## marks the start of existence of a non-owning location
+    # future direction: remove this distinction and perform all related decision
+    # making (e.g., injecting destructors) requiring knowledge of locations'
+    # ownership in ``mirgen``. There's only going to be the ``Def`` kind
+    mnkDefUnpack ## intermediate hack required by destructor injection. Don't
+                 ## use
+    mnkBind      ## introduces an alias that may only be used for read access.
+                 ## The source expression must not be empty
+    mnkBindMut   ## introduces an alias that may be used for write access.
+                 ## The source expression must not be empty
+
+    mnkFastAsgn ## assignment that cannot be rewritten into copy, move, or
+                ## hook call
+    # future direction: same as with DefCursor, remove FastAsgn
+    mnkAsgn     ## normal assignment; the destination might store a value
+                ## already. Whether the source is copied or moved depends
+                ## on the expression
+    mnkInit     ## similar to ``mnkAsgn``, but with the guarantee that the
+                ## destination is empty
```

Comment:
I think "empty" is a bad qualifier here, but something like "not initialized" is better.

My guess with this type is that you want to handle the initialization case where destruction could be avoided, but it should not be a guarantee that it will be "empty" as in 0-init.

Reply:
"empty" is not a guarantee about the bit pattern of the location, only that the location doesn't store a value (in the logical sense). Nonetheless, "not initialized" is indeed more clear.

Reply:
I've changed the comment to use "not initialized" rather than "empty".